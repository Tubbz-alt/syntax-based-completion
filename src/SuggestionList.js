import React from "react";
import PropTypes from "prop-types";

import LoadingSpinner from "./LoadingSpinner";

export default class SuggestionList extends React.Component {
  static propTypes = {
    loading: PropTypes.bool,
    suggestions: PropTypes.array,
    active: PropTypes.number,
    selected: PropTypes.string,
    onSelect: PropTypes.func.isRequired,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func
  };

  static defaultProps = {
    suggestions: []
  };

  onClickSuggestion = suggestion => e => {
    e.stopPropagation();
    e.preventDefault();
    this.props.onSelect(suggestion);
  };

  render() {
    return (
      <ul
        className="query-suggestion-list"
        onMouseEnter={this.props.onMouseEnter}
        onMouseLeave={this.props.onMouseLeave}
      >
        {this.props.loading ? (
          <div className="query-suggestion-loading">
            <LoadingSpinner />
          </div>
        ) : (
          this.props.suggestions.map((suggestion, index) => {
            const active = index === this.props.active;
            const selected = suggestion === this.props.selected;
            return (
              <li
                key={suggestion}
                className={`query-suggestion-item ${active ? "active" : ""} ${
                  selected ? "selected" : ""
                }`}
                onClick={this.onClickSuggestion(suggestion)}
              >
                {suggestion}
              </li>
            );
          })
        )}
      </ul>
    );
  }
}
