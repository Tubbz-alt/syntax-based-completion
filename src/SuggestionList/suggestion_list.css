
.SuggestionList {
    list-style-type: none;
    margin: 0;
    margin-top: -1px;
    padding: 0;
    position: absolute;
    width: 100%;
    z-index: 1;

    border: 1px solid #ddd;
    border-top: 0;
    background-color: #fff;
    box-shadow: 0 2px 6px -5px black;

    max-height: 60vh;
    overflow: auto;
}

.SuggestionList__loading {
    display: block;
    text-align: center;
}

.SuggestionList__item,
.SuggestionList__item--active,
.SuggestionList__item--selected {
    display: block;
    padding: .2rem .5rem;
    font-size: 1rem;
    text-align: left;
    font-family: inherit;
    font-size: inherit;
    font-stretch: inherit;
    border: 1px solid #fff;
    color: #666;
}

.SuggestionList__item--active,
.SuggestionList__item:hover {
    border: 1px solid #ddd;
}

.SuggestionList__item--selected {
    background: #f0f0f0;
}
