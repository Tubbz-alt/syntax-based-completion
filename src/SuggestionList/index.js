import SuggestionList from "./suggestion_list";
export default SuggestionList;
