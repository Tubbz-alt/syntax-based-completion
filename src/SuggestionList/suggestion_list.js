import React from "react";
import PropTypes from "prop-types";

import "./suggestion_list.css";

import LoadingSpinner from "../LoadingSpinner";

export default class SuggestionList extends React.Component {
  static propTypes = {
    loading: PropTypes.bool,
    suggestions: PropTypes.array,
    active: PropTypes.number,
    selected: PropTypes.string,
    onSelect: PropTypes.func.isRequired,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func
  };

  static defaultProps = {
    suggestions: []
  };

  onClickSuggestion = suggestion => e => {
    e.stopPropagation();
    e.preventDefault();
    this.props.onSelect(suggestion);
  };

  renderSuggestion = (suggestion, index) => {
    let active = index === this.props.active;
    let selected = suggestion === this.props.selected;
    let itemClassName = `SuggestionList__item${
      active ? "--active" : selected ? "--selected" : ""
    }`;
    return (
      <li
        key={suggestion}
        className={itemClassName}
        onClick={this.onClickSuggestion(suggestion)}
      >
        {suggestion}
      </li>
    );
  };

  render() {
    return (
      <ul
        className="SuggestionList"
        onMouseEnter={this.props.onMouseEnter}
        onMouseLeave={this.props.onMouseLeave}
      >
        {this.props.loading ? (
          <div className="SuggestionList__loading">
            <LoadingSpinner />
          </div>
        ) : (
          this.props.suggestions.map(this.renderSuggestion)
        )}
      </ul>
    );
  }
}
