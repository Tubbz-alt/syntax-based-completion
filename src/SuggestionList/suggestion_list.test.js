import React from "react";
import { shallow } from "enzyme";

import SuggestionList from "./suggestion_list";

describe("<SuggestionList>", () => {
  describe("shallow render", () => {
    it("loading", () => {
      const wrapper = shallow(<SuggestionList loading />);

      expect(wrapper.find("LoadingSpinner")).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    });

    it("with suggestions", () => {
      const wrapper = shallow(<SuggestionList suggestions={["Foo", "Bar"]} />);
      expect(wrapper.find("LoadingSpinner")).toHaveLength(0);
      expect(wrapper.find("li.SuggestionList__item")).toHaveLength(2);
      expect(wrapper).toMatchSnapshot();
    });

    it("with active suggestions", () => {
      const wrapper = shallow(
        <SuggestionList suggestions={["Foo", "Bar"]} active={1} />
      );
      expect(wrapper.find("LoadingSpinner")).toHaveLength(0);
      expect(wrapper.find(".SuggestionList__item")).toHaveLength(1);
      expect(wrapper.find(".SuggestionList__item--active")).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    });

    it("with active and selected suggestions", () => {
      const wrapper = shallow(
        <SuggestionList
          suggestions={["Foo", "Bar"]}
          active={1}
          selected={"Foo"}
        />
      );
      expect(wrapper.find("LoadingSpinner")).toHaveLength(0);
      expect(wrapper.find(".SuggestionList__item")).toHaveLength(0);
      expect(wrapper.find(".SuggestionList__item--active")).toHaveLength(1);
      expect(wrapper.find(".SuggestionList__item--selected")).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("onClickSuggestion", () => {
    it("prevent default click behavior before calling change", () => {
      const mockedOnSelect = jest.fn();
      const mockedEvent = {
        stopPropagation: jest.fn(),
        preventDefault: jest.fn()
      };
      const wrapper = shallow(
        <SuggestionList
          suggestions={["Foo", "Bar"]}
          onSelect={mockedOnSelect}
        />
      );

      wrapper.instance().onClickSuggestion("Bar")(mockedEvent);

      expect(mockedEvent.stopPropagation).toHaveBeenCalled();
      expect(mockedEvent.preventDefault).toHaveBeenCalled();
      expect(mockedOnSelect).toHaveBeenCalledWith("Bar");
    });
  });
});
