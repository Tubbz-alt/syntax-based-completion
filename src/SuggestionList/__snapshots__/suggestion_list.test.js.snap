// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SuggestionList> shallow render loading 1`] = `
<ul
  className="SuggestionList"
>
  <div
    className="SuggestionList__loading"
  >
    <LoadingSpinner />
  </div>
</ul>
`;

exports[`<SuggestionList> shallow render with active and selected suggestions 1`] = `
<ul
  className="SuggestionList"
>
  <li
    className="SuggestionList__item--selected"
    key="Foo"
    onClick={[Function]}
  >
    Foo
  </li>
  <li
    className="SuggestionList__item--active"
    key="Bar"
    onClick={[Function]}
  >
    Bar
  </li>
</ul>
`;

exports[`<SuggestionList> shallow render with active suggestions 1`] = `
<ul
  className="SuggestionList"
>
  <li
    className="SuggestionList__item"
    key="Foo"
    onClick={[Function]}
  >
    Foo
  </li>
  <li
    className="SuggestionList__item--active"
    key="Bar"
    onClick={[Function]}
  >
    Bar
  </li>
</ul>
`;

exports[`<SuggestionList> shallow render with suggestions 1`] = `
<ul
  className="SuggestionList"
>
  <li
    className="SuggestionList__item"
    key="Foo"
    onClick={[Function]}
  >
    Foo
  </li>
  <li
    className="SuggestionList__item"
    key="Bar"
    onClick={[Function]}
  >
    Bar
  </li>
</ul>
`;
